### PR TITLE
Create slice STDCM and simulation#5272

### LINKED ISSUE #5272: Create slice STDCM and simulation
- [x] add stdcm slice
- [x] add operationalSutdy (simulation) slice
- [x] create common reducers for (stdcm and simulation)
- [x] add tests for operationalStudy state
- [x] add tests for stdcm state
- [x] add selectors for stdcm/simulation slices
- [x] add context to inject the right actions between stdcm and operationalStudy
- [x] Check if updateInfraID trigger getInfraByIdSwitchTypes

### PR DESCRIPTION
**Pull request by kmer2016**
Closes #5272

### Description
This marks the initial phase of decoupling the simulation functionality from `stdcm`. We are introducing two separate slices to handle each aspect independently. While we haven't completely replaced `osrdconf` at this stage, we've integrated it alongside our two new state slices. The end goal is to gradually deplete `osrdconf` and fully transition to the new slices.

### Key Highlights:
1. **Common Reducers Object** : We've created a shared reducers object that both slices will utilize, promoting code modularity and maintainability.
2. **updateSwitchTypes Action Refactoring** : Instead of triggering the action each time the infrastructure ID changes, which was leading to unnecessary calls, we've now streamlined it. We directly invoke the RTK query to fetch switch types precisely when and where they are needed.
3. osrdConf context :  We've implemented a react context to specify from which slice (stdcmconf or simulationconf), a component have to call a osrdconf actions 
4. **Testing** : We've added tests for both of the new slices.

### How to test ?

1. Start the application.
2. Inspect the Redux store to verify the presence of two new fields: `stdcmconf` and `simulationconf` in the state.
3. Navigate to the stdcm view and examine the `stdcmconf` state to confirm the existence of an `infraID`. In contrast, inspect the `simulationconf` state to verify the absence of infraID.

### Next step
Find a way to test the interactions between `updateInfraID`, `getInfraByIdSwitchTypes`. This involves mocking the RTK query for efficient testing. Following resources can help : 
  - https://dev.to/ifeanyichima/-testing-components-with-a-request-for-rtk-query-using-msw-and-react-testing-library-5a8n
  - https://remarkablemark.medium.com/how-to-test-redux-toolkit-query-with-jest-and-react-testing-library-cc02c0528fcf
  - https://gustavocd.dev/posts/testing-rtk-query-with-msw/

Additionally, we plan to replace occurrences of osrdConf.simulationConf with simulationConf and osrdConf.stdcmConf with stdcmConf in the store to maintain consistency and clarity (as noted in issues #5273 and #5274).

### CODE REVIEW

File: front/src/Store.ts [hunk 1]
```diff
@@ -20,13 +22,25 @@ const reduxDevToolsOptions: Config = {
...
+
 const middlewares: Middleware[] = [thunk, osrdEditoastApi.middleware];
 
 const store = configureStore({
   reducer: persistedReducer,
   devTools: reduxDevToolsOptions,
   middleware: (getDefaultMiddleware) =>
-    getDefaultMiddleware({ serializableCheck: false }).concat(...middlewares),
+    getDefaultMiddleware({ serializableCheck: false })
+      // As mentionned in the Doc(https://redux-toolkit.js.org/api/createListenerMiddleware), Since this can receive actions with functions inside,
```
Comment (clarani) [inline front/src/Store.ts]: Add a new line and remove capital letter. Nice comment :rocket: 
```suggestion
      // As mentionned in the Doc(https://redux-toolkit.js.org/api/createListenerMiddleware), 
      // since this can receive actions with functions inside,
```

File: front/src/Store.ts [hunk 2]
```diff
@@ -20,13 +22,25 @@ const reduxDevToolsOptions: Config = {
   },
 };
 
+const buildListenerMiddleware = () => {
+  const listener = createListenerMiddleware();
+  registerSimulationUpdateInfraIDListeners(listener);
+  registerStdcmUpdateInfraIDListeners(listener);
+
+  return listener;
+};
```
Comment (clarani) [inline front/src/Store.ts]: I think that switchTypes is only used in the editor, so maybe we should move the call to fetch the switch type directly in the editor, rather call it every time we change infra in the other applications. If we do this, we don't need this listener anymore
```suggestion
const buildListenerMiddleware = () => {
  const listener = createListenerMiddleware();
  registerSimulationUpdateInfraIDListeners(listener);
  registerStdcmUpdateInfraIDListeners(listener);

  return listener;
};
```

File: front/src/reducers/index.ts [hunk 3]
```diff
@@ -88,6 +99,8 @@ export const rootInitialState: RootState = {
   map: mapInitialState,
   editor: editorInitialState,
   main: mainInitialState,
+  stdcmconf: stdcmConfInitialState,
+  simulationconf: simulationConfInitialState,
```
Comment (clarani) [inline front/src/reducers/index.ts]:
```suggestion
  stdcmConf: stdcmConfInitialState,
  simulationConf: simulationConfInitialState,
```

File: front/src/reducers/osrdconf2/stdcmConf/stdcmConfReducers.spec.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    store.dispatch(stdcmConfSliceActions.updateStdcmMode(newStdcmMode));
+
+    const state = store.getState().stdcmconf;
+    expect(state.stdcmMode).toBe(newStdcmMode);
+  });
+
+  it('should handle updateStdcmStandardAllowance', () => {
+    const store = createStore({
+      standardStdcmAllowance: testDataBuilder.buildTimeStandardAllowance(10),
+    });
```
Comment (clarani) [inline front/src/reducers/osrdconf2/stdcmConf/stdcmConfReducers.spec.ts:70]: Should we check that the allowance is correctly initialized first ?
```suggestion
    });
    expect(state.standardStdcmAllowance).toBe(timeStandardAllowance);
```

File: front/src/reducers/osrdconf2/stdcmConf/stdcmConfReducers.spec.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    const newStdcmMode = testDataBuilder.buildPercentageStandardAllowance(5);
```
Comment (clarani) [inline front/src/reducers/osrdconf2/stdcmConf/stdcmConfReducers.spec.ts]:
```suggestion
    const newStandardAllowance = testDataBuilder.buildPercentageStandardAllowance(5);
```

File: front/src/reducers/osrdconf2/stdcmConf/stdcmConfReducers.spec.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    store.dispatch(stdcmConfSliceActions.updateMaximumRunTime(newMaximumRunTime));
+
+    const state = store.getState().stdcmconf;
+    expect(state.maximumRunTime).toBe(newMaximumRunTime);
+  });
+
+  it('should handle updateStdcmMode', () => {
+    const store = createStore({
+      stdcmMode: STDCM_MODES.byDestination,
+    });
```
Comment (clarani) [inline front/src/reducers/osrdconf2/stdcmConf/stdcmConfReducers.spec.ts:54]:
```suggestion
    });
    expect(state.stdcmMode).toBe(STDCM_MODES.byDestination);
```

File: front/src/reducers/osrdconf2/simulationConf/index.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  addCommonOsrdConfMatchers,
+  registerUpdateInfraIDListener,
+} from '../common';
+
+export const simulationConfInitialState = defaultCommonConf;
+
+export const simulationConfSlice = createSlice({
+  name: 'simulationconf',
+  initialState: simulationConfInitialState,
+  reducers: { ...buildCommonOsrdConfReducers<OsrdConfState>() },
```
Comment (clarani) [inline front/src/reducers/osrdconf2/simulationConf/index.ts]: Can you harmonize the names ? defaultCommonConf, buildCommonConfReducers, buildCommonConfSelectors, testCommonConfReducers 
```suggestion
  reducers: { ...buildCommonConfReducers<OsrdConfState>() },
```

File: front/src/reducers/osrdconf2/common/index.ts [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+    updateOriginDate(state: Draft<S>, action: PayloadAction<S['originDate']>) {
+      state.originDate = action.payload;
+    },
+    updateOriginUpperBoundDate(state: Draft<S>, action: PayloadAction<S['originUpperBoundDate']>) {
+      state.originUpperBoundDate = action.payload;
+    },
+    replaceVias(state: Draft<S>, action: PayloadAction<S['vias']>) {
+      state.vias = action.payload;
+    },
+    updateVias(state: Draft<S>, action: PayloadAction<PointOnMap>) {
```
Comment (clarani) [inline front/src/reducers/osrdconf2/common/index.ts:213]: Can you left a comment here to rename this action  "addVia" ? 
```suggestion
    addVia(state: Draft<S>, action: PayloadAction<PointOnMap>) {
```

File: front/src/reducers/osrdconf2/common/index.ts
Comment (clarani) [inline front/src/reducers/osrdconf2/common/index.ts:1]: Some of these props are only used in stdcm or simulation, we will need to clean this up later :)

File: front/src/applications/operationalStudies/consts.ts [hunk 9]
```diff
@@ -180,7 +180,7 @@ export interface OsrdConfState {
   departureTime: string;
   destination?: PointOnMap;
   vias: PointOnMap[];
-  suggeredVias: Path['steps'];
+  suggeredVias: Path['steps'] | PointOnMap[];
```
Comment (clarani) [inline front/src/applications/operationalStudies/consts.ts:183]: Why did you add this ? Is it for the tests ?
```suggestion
  suggeredVias: Path['steps'] | PointOnMap[];
```

File: front/src/reducers/osrdconf2/common/selectors.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
+import { RootState } from 'reducers';
+import { simulationConfSliceType } from '../simulationConf';
+import { stdcmConfSliceType } from '../stdcmConf';
+
+const buildSelectors = (slice: simulationConfSliceType | stdcmConfSliceType) => ({
```
Comment (clarani) [inline front/src/reducers/osrdconf2/common/selectors.ts]:
```suggestion
const buildCommonConfSelectors = (slice: simulationConfSliceType | stdcmConfSliceType) => ({
```

File: front/src/reducers/osrdconf2/common/testUtils.ts [hunk 11]
```diff
@@ -0,0 +1,681 @@
...
+      {
+        value: 'test',
+        begin: 1,
+        end: 2,
+      },
+    ],
+  };
+}
+
+const testCommonReducers = (slice: simulationConfSliceType | stdcmConfSliceType) => {
```
Comment (clarani) [inline front/src/reducers/osrdconf2/common/testUtils.ts]:
```suggestion
const testCommonConfReducers = (slice: simulationConfSliceType | stdcmConfSliceType) => {
```

File: front/src/reducers/osrdconf2/common/index.ts [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+    ) {
+      state.trainScheduleIDsToModify = action.payload;
+    },
+  };
+}
+
+/**
+ *
+ */
+export function addCommonOsrdConfMatchers<S extends OsrdConfState>(
```
Comment (clarani) [inline front/src/reducers/osrdconf2/common/index.ts]:
```suggestion
export function addCommonConfMatchers<S extends OsrdConfState>(
```

**Review by clarani — COMMENTED** [review]
Thanks for this PR :pray:

File: front/src/reducers/osrdconf2/common/testUtils.ts [hunk 13]
```diff
@@ -0,0 +1,681 @@
...
+) {
+  return createStoreWithoutMiddleware({
+    [slice.name]: { ...defaultCommonConf, ...initialStateExtra },
+  });
+}
+
+function osrdCommonConfTestDataBuilder() {
+  return {
+    buildEngineeringAllowance: (): Allowance => ({
+      allowance_type: 'engineering',
```
Comment (anisometropie) [inline front/src/reducers/osrdconf2/common/testUtils.ts]: we can’t have that much test data hard coded in function body. 
turn that into json files or some factory functions

File: front/src/Store.ts [hunk 14]
```diff
@@ -20,13 +22,25 @@ const reduxDevToolsOptions: Config = {
   },
 };
 
+const buildListenerMiddleware = () => {
```
Comment (anisometropie) [inline front/src/Store.ts]: we need to merge that with the config I made in 
https://github.com/osrd-project/osrd/pull/4876/files#diff-e134333a314faf912a3474872afce536f49db9ab4f4dac0f9090cb10dbfa67e3

File: front/src/reducers/osrdconf2/common/selectors.ts (hunk 10 above)
Comment (anisometropie) [inline front/src/reducers/osrdconf2/common/selectors.ts]: Can’t we use `makeSubSelector`?

File: front/src/reducers/osrdconf2/common/testUtils.ts [hunk 15]
```diff
@@ -0,0 +1,681 @@
...
+        end: 2,
+      },
+    ],
+  };
+}
+
+const testCommonReducers = (slice: simulationConfSliceType | stdcmConfSliceType) => {
+  const testDataBuilder = osrdCommonConfTestDataBuilder();
+
+  it('should handle updateName', () => {
```
Comment (anisometropie) [inline front/src/reducers/osrdconf2/common/testUtils.ts]: why do you have unit test inside a file called testUtils ?

**Review by anisometropie — CHANGES_REQUESTED** [review]

File: front/src/applications/operationalStudies/consts.ts (hunk 9 above)
Comment (kmer2016) [reply front/src/applications/operationalStudies/consts.ts:183]: because I saw that in `adjustConfWithTrainToModify.ts` file, when we have `dispatch(updateSuggeredVias(compact(path.steps.map((via) => convertStepToPointOnMap(via)))));` we pass an array of type PointOnMap[] in updateSuggeredVias...

Comment (anisometropie) [reply front/src/applications/operationalStudies/consts.ts:183]: No `convertStepToPointOnMap` returns `PointOnMap`

File: front/src/applications/editor/tools/switchEdition/types.ts
Comment (clarani) [inline front/src/applications/editor/tools/switchEdition/types.ts:1]: :raised_hands:

File: front/src/reducers/osrdconf2/common/index.ts [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+      state.originDate = action.payload;
+    },
+    updateOriginUpperBoundDate(state: Draft<S>, action: PayloadAction<S['originUpperBoundDate']>) {
+      state.originUpperBoundDate = action.payload;
+    },
+    replaceVias(state: Draft<S>, action: PayloadAction<S['vias']>) {
+      state.vias = action.payload;
+    },
+    // TODO rename updateVias -> addVias
+    addVias(state: Draft<S>, action: PayloadAction<PointOnMap>) {
```
Comment (clarani) [inline front/src/reducers/osrdconf2/common/index.ts]: Hum if we rename it here, then we also need to change the osrdconf reducer, otherwise this will break when we will switch to this store slice.
So 2 possibilities here : 
- rename also in osrdconf now, rename all the usages (only 1-2 I think) and remove the comment line 212
- rename back to updateVias in this slice, and leave the comment so we'll do the rename later

Personally, I think we can do the rename later, but the choice is yours. Also, as you can add only 1 via at the time, I think addVia without S would be a better name.

```suggestion
    addVias(state: Draft<S>, action: PayloadAction<PointOnMap>) {
```

File: front/src/reducers/osrdconf2/common/testUtils.ts [hunk 17]
```diff
@@ -0,0 +1,674 @@
...
+  it('should handle updateTrainScheduleIDsToModify', () => {
+    const store = createStore(slice);
+    const newTrainScheduleIDsToModify = [10, 2];
+    store.dispatch(slice.actions.updateTrainScheduleIDsToModify(newTrainScheduleIDsToModify));
+    const state = store.getState()[slice.name];
+    expect(state.trainScheduleIDsToModify).toStrictEqual(newTrainScheduleIDsToModify);
+  });
+};
+
+function commonConfTestDataBuilder() {
```
Comment (clarani) [inline front/src/reducers/osrdconf2/common/testUtils.ts]: Don't you want to move this function to the top of the file ? Thus you can remove the eslint warning

File: front/src/reducers/osrdconf2/common/testUtils.ts [hunk 18]
```diff
@@ -0,0 +1,674 @@
...
+
+  it('should handle replaceVias', () => {
+    const store = createStore(slice);
+    const newVias: PointOnMap[] = [testDataBuilder.buildPointOnMap()];
+    store.dispatch(slice.actions.replaceVias(newVias));
+    const state = store.getState()[slice.name];
+    expect(state.vias).toBe(newVias);
+  });
+
+  it('should handle updateVias', () => {
```
Comment (clarani) [inline front/src/reducers/osrdconf2/common/testUtils.ts]: If you rename updateVias -> addVia, don't forget to change the test description :)
```suggestion
  it('should handle addVias', () => {
```

**Review by clarani — COMMENTED** [review]
Very nice job, I left some last comments

File: front/src/reducers/osrdconf2/common/index.ts (hunk 16 above)
Comment (kmer2016) [reply front/src/reducers/osrdconf2/common/index.ts]: For now, I keep updateVias, I will rename it when I will switch to store slice.

**Review by clarani — APPROVED** [review]
LGTM and tested :heavy_check_mark:

**Review by anisometropie — APPROVED** [review]
